pkg/util/containers/v2/metrics/docker/collector.go

```diff
@@ -80,6 +92,59 @@ func (d *dockerCollector) GetContainerNetworkStats(containerID string, cacheVali
 	return convertNetworkStats(stats.Networks), nil
 }
 
+// GetContainerIDForPID returns the container ID for given PID
+func (d *dockerCollector) GetContainerIDForPID(pid int, cacheValidity time.Duration) (string, error) {
+	currentTime := time.Now()
+	strPid := strconv.Itoa(pid)
+
+	cID, found, _ := d.pidCache.Get(currentTime, strPid, cacheValidity)
+	if found {
+		return cID.(string), nil
+	}
+
+	// If we've done a full refresh within cacheValidity, we do not trigger another full refresh
+	_, found, _ = d.pidCache.Get(currentTime, pidCacheFullRefreshKey, cacheValidity)
+	if found {
+		return "", nil
+	}
```

Comment:
This is an obtuse way to check if enough time has passed since the last full refresh (it took me a fair few reads to understand what this does). I think an explicit check for time elapsed is much clearer.

Reply:
I'd require to store `refreshTimestamp`, `refreshErr` and implement the logic based on `cacheValidity` and know when the cache has been `GC-ed` or not.

Using the cache itself seems like a better solution both in terms of consistency and regarding DRY. I agree it can be better documented though.